wordpress-mobile/WordPress-Android
# Detekt Integration - Phase.1

External Links: [Detekt](https://github.com/detekt/detekt) - [Detekt Documentation](https://detekt.github.io/detekt/)
Internal Link: `paqN3M-eg-p2` (Detekt Integration)
Related PR: [Showcase](https://github.com/wordpress-mobile/WordPress-Android/pull/13290)
Next Phases: [Phase.2](https://github.com/wordpress-mobile/WordPress-Android/issues/14011) - [Phase.3](https://github.com/wordpress-mobile/WordPress-Android/issues/14012)

This PR adds `Detekt` to the `WordPress` repo. Detekt is a static code analysis tool for the Kotlin programming language. Detekt does for Kotlin what CheckStyle does for Java, only so much more.

To test (🧪 ):
- Make sure to read the external and internal links mention above to get an overall idea on what `Detekt` has to offer, how it can be configured and utilised.
- Pull `detekt-integration` on a local branch.
- Run `./gradlew WordPress:detekt` and verify that the build is successful.
- Delete the `config/detekt/baseline.xml` file, comment out the `Detekt's` baseline configuration within the `WordPress/build.gradle` file and run `./gradlew WordPress:detekt` again. Verify that the build fails, reporting `880` issues.
- Install the `detekt` IDE plugin and configure it following the instructions within the `coding-style.md` readme file. Verify that everything works and that the IDE reports `Detekt` violations as well.
- Locate the `Detekt` related `Circle CI` job and verify correctness.

Questions (❓):
- For @loremattei: `Detekt` was configured on the root `build.gradle` file, within the `allprojects { ... }` scope. However, I have noticed that the `subprojects { ... }` is utilised as well for static analysis tools like `KtLint`. Do I need to add an additional `task` for `Detekt` within the `subprojects { ... }`?
- For @loremattei: The exist a task within the WordPress `build.gradle` file, called `violationCommentsToGitHub`. This task has a `violations` configuration for `CheckStyle`. Do I need to add something similar for `Detekt`?

**Merge Instructions** (⚠️):
1. Make sure 2 reviewers have approved the PR (@loremattei and @malinajirka).
1. Remove the `Not Ready for Merge` label.
1. Merge this PR.

PR submission checklist:
